deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,698 @@
+from lockfile import LockFile
+from deepdrivemd.data.stream.aggregator_reader import (
+    Streams,
+    StreamVariable,
+    StreamContactMapVariable,
+    StreamScalarVariable,
+)
+
+import cupy as cp
+from cuml import DBSCAN as DBSCAN
+
+from deepdrivemd.models.keras_cvae.model import CVAE
+from simtk.openmm.app.pdbfile import PDBFile
+
+import adios2
+
+
+def clear_gpu():
+    K.clear_session()
+    try:
+        device = int(os.environ["CUDA_VISIBLE_DEVICES"])
+        print("device = ", device)
+        cuda.select_device(device)
+        cuda.close()
+    except Exception as e:
+        print(e)
+
+
+def build_model(cfg: OutlierDetectionConfig, model_path: str):
+    cvae = CVAE(
+        image_size=cfg.final_shape,
+        channels=cfg.final_shape[-1],
+        conv_layers=cfg.conv_layers,
+        feature_maps=cfg.conv_filters,
+        filter_shapes=cfg.conv_filter_shapes,
+        strides=cfg.conv_strides,
+        dense_layers=cfg.dense_layers,
+        dense_neurons=cfg.dense_neurons,
+        dense_dropouts=cfg.dense_dropouts,
+        latent_dim=cfg.latent_dim,
+    )
+    cvae.load(model_path)
+    return cvae
+
+
+def wait_for_model(cfg: OutlierDetectionConfig) -> str:
+    """Wait for the trained model to be published by machine learning pipeline.
+
+    Returns
+    -------
+    str
+         path to the model
+    """
+
+    while True:
+        if os.path.exists(cfg.best_model):
+            break
+        print(f"No model {cfg.best_model}, sleeping")
+        sys.stdout.flush()
+        time.sleep(cfg.timeout2)
+    return cfg.best_model
+
+
+def wait_for_input(cfg: OutlierDetectionConfig) -> List[str]:
+    """Wait for enough data to be produced by simulations.
+
+    Returns
+    -------
+    List[str]
+         a list of aggregated bp file
+    """
+    while True:
+        bpfiles = glob.glob(str(cfg.agg_dir / "*/*/agg.bp"))
+        if len(bpfiles) == cfg.num_agg:
+            break
+        print(f"Waiting for {cfg.num_agg} agg.bp files")
+        time.sleep(cfg.timeout1)
+
+    print(f"bpfiles = {bpfiles}")
+
+    # Wait for enough time steps in each bp file
+    while True:
+        enough = True
+        for bp in bpfiles:
+            com = f"bpls {bp}"
+            a = subprocess.getstatusoutput(com)
+            if a[0] != 0:
+                enough = False
+                print(f"Waiting, a = {a}, {bp}")
+                break
+            try:
+                steps = int(a[1].split("\n")[0].split("*")[0].split(" ")[-1])
+            except Exception as e:
+                print("Exception ", e)
+                steps = 0
+                enough = False
+            if steps < cfg.min_step_increment:
+                enough = False
+                print(f"Waiting, steps = {steps}, {bp}")
+                break
+        if enough:
+            break
+        else:
+            time.sleep(cfg.timeout2)
+
+    return bpfiles
+
+
+def dirs(cfg: OutlierDetectionConfig) -> Tuple[str, str, str]:
```

Comment:
Refactor:

```
tmp_dir = cfg.output_path / "tmp"
published_dir = cfg.output_path / "published_outliers"
tmp_dir.mkdir(exists_ok=True)
published_dir.mkdir(exists_ok=True)
return cfg.output_path, tmp_dir, published_dir
```

What does `top_dir` refer to? It looks like it is not used in the rest of the code and can be removed. Then we could simply return `tmp_dir, published_dir`. 

Also, in the docstring can you please add some comments explaining what data is written to each of the tmp and published_outliers dirs?

Reply:
Done.